include/deal.II/base/hdf5.h

```diff
@@ -0,0 +1,1073 @@
+
+    /**
+     * HDF5 dataspace identifier.
+     */
+    std::shared_ptr<hid_t> dataspace;
+
+    /**
+     * Total number of elements in the dataset.
+     */
+    unsigned int size;
+
+    /**
+     * If query_io_mode is set to true, then after every read/write operation
+     * the HDF5 deal.ii interface calls the routines
+     * [H5Pget_mpio_actual_io_mode()](https://support.hdfgroup.org/HDF5/doc/RM/RM_H5P.html#Property-GetMpioActualIoMode)
+     * and
+     * [H5Pget_mpio_no_collective_cause()](https://support.hdfgroup.org/HDF5/doc/RM/RM_H5P.html#Property-GetMpioNoCollectiveCause).
+     * The results are stored in io_mode, local_no_collective_cause and
+     * global_no_collective_cause.
+     */
+    bool query_io_mode;
+
+    /**
+     * I/O mode that was performed on the last parallel I/O call.
+     */
+    H5D_mpio_actual_io_mode_t io_mode;
+
+    /**
+     * Local causes that broke collective I/O on the
+     * last parallel I/O call. See <a
+     * href="https://support.hdfgroup.org/HDF5/doc/RM/RM_H5P.html#Property-GetMpioNoCollectiveCause">H5Pget_mpio_no_collective_cause</a>.
+     */
+    uint32_t local_no_collective_cause;
+
+    /**
+     * Global causes that broke collective I/O on the
+     * last parallel I/O call. See <a
+     * href="https://support.hdfgroup.org/HDF5/doc/RM/RM_H5P.html#Property-GetMpioNoCollectiveCause">H5Pget_mpio_no_collective_cause</a>.
+     */
+    uint32_t global_no_collective_cause;
+  };
+
+  /**
+   * This class implements an HDF5 Group
+   *
+   * @author Daniel Garcia-Sanchez, 2018
+   */
+  class Group : public HDF5Object
+  {
+  protected:
+    /**
+     * Group access mode
+     */
+    enum class GroupAccessMode
+    {
+      /**
+       * Opens an existing group
+       */
+      open,
+      /**
+       * Creates a new group
+       */
+      create
+    };
+    /**
+     * This constructor creates or opens a group depending on the value of
+     * @p mode. The group will be placed insided the group @p parent_group. The
+     * parameter @p mpi defines if the the I/O operations are serial or
+     * parallel. This is an internal constructor, the functions open_group() and
+     * create_group() should be used to open or create a group.
+     */
+    Group(const std::string &   name,
+          const Group &         parent_group,
+          const bool            mpi,
+          const GroupAccessMode mode);
+
+    /**
+     * Internal constructor used by File. The constructor sets the protected
+     * const members of HDF5Group: @p name and @p mpi. It does not create or
+     * open a Group.
+     */
+    Group(const std::string &name, const bool mpi);
+
+  public:
+    /**
+     * Opens a group.
+     */
+    Group
+    open_group(const std::string &name);
+
+    /**
+     * Creates a group.
+     */
+    Group
+    create_group(const std::string &name);
+
+    /**
+     * Opens a dataset.
+     */
+    DataSet
+    open_dataset(const std::string &name);
```

Comment:
For the two `open_*` functions of this class, could they be `const`? That is, does opening a group change the state of the object?

Reply:
I made these functions const. I'll push all the commits when I finish.

The functions `read_*` and `write_*` of DataSet "could" be const. See my comment below.